### DISCUSSION TITLE
[Question] How to retrieve metadata of document using spring AI? #4173

### DISCUSSION THREAD

**Discussion by vineetaparodkar** [spring-projects/spring-ai#678] [Q&A]
https://github.com/orgs/qdrant/discussions/4173 

How to retrieve metadata of document using Qdrant and springAI? ?Basically need to show metadata like filename, chunkname etc. Following code snippet describes starting point for retrieving based on query. How to modify this to get metadata of documents too i.e. we need metadata for RAG QnA application which will be pdf names which were provided as context.

Application code - https://github.com/vineetaparodkar/rag-with-openai-spring-ai.git 

```
public String askLlm(String query) {
        List<Document> documentList = vectorStore.similaritySearch(query);

        String systemMessageTemplate = """
                Answer the question in JSON format but do not add ```json``` , based solely on the provided CONTEXT.
                If the answer is not found in the context, respond 'I don't know'.
                CONTEXT:
                     {CONTEXT}
                """;

        Message systemMessage = new SystemPromptTemplate(systemMessageTemplate)
                .createMessage(Map.of("CONTEXT", documentList));
        UserMessage userMessage = new UserMessage(query);
        Prompt prompt = new Prompt(List.of(systemMessage, userMessage));
        OpenAiApi aiApi = new OpenAiApi(apiKey);
        OpenAiChatOptions openAiChatOptions = OpenAiChatOptions.builder()
                .withModel(model)
                .withTemperature(temperature)
                .withMaxTokens(maxToken)
                .build();
        OpenAiChatClient openAiChatClient = new OpenAiChatClient(aiApi, openAiChatOptions);
        ChatResponse response = openAiChatClient.call(prompt);
        String responseContent = response.getResult().getOutput().getContent();
        return responseContent;

    }```

**Comment by f3rnandomoreno** [comment]
When you follow the ETL pipeline, 

https://docs.spring.io/spring-ai/reference/api/etl-pipeline.html

you first extract the data with a DocumentReader, this extract depend of the class you use include the file name in every document in the metadata.
If not in the second step, in the transformation you can enrich every document with metadata, so if you extract the documents from a file after that you include in all Documents of the list the metadata "filename" with the file name.

After you have Transform the Documents you load all documents in the VectorStore. So they have already all metadata related with the filename.

Then after this, when you retrieve the document from the vector store you will have this information in the metadata of the Document.
